plan.go

```diff
@@ -0,0 +1,92 @@
+package tfe
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"time"
+)
+
+// PlanStatus represents a plan state.
+type PlanStatus string
+
+//List all available plan statuses.
+const (
+	PlanCanceled   PlanStatus = "canceled"
+	PlanCreated    PlanStatus = "created"
+	PlanErrored    PlanStatus = "errored"
+	PlanFinished   PlanStatus = "finished"
+	PlanMFAWaiting PlanStatus = "mfa_waiting"
+	PlanPending    PlanStatus = "pending"
+	PlanQueued     PlanStatus = "queued"
+	PlanRunning    PlanStatus = "running"
+)
+
+// Plans handles communication with the plan related methods of the Terraform
+// Enterprise API.
+//
+// TFE API docs: https://www.terraform.io/docs/enterprise/api/plan.html
+type Plans struct {
+	client *Client
+}
+
+// Plan represents a Terraform Enterprise plan.
+type Plan struct {
+	ID               string                `jsonapi:"primary,plans"`
+	HasChanges       bool                  `jsonapi:"attr,has-changes"`
+	LogReadURL       string                `jsonapi:"attr,log-read-url"`
+	Status           PlanStatus            `jsonapi:"attr,status"`
+	StatusTimestamps *PlanStatusTimestamps `jsonapi:"attr,status-timestamps"`
+}
+
+// PlanStatusTimestamps holds the timestamps for individual plan statuses.
+type PlanStatusTimestamps struct {
+	CanceledAt   time.Time `json:"canceled-at"`
+	CreatedAt    time.Time `json:"created-at"`
+	ErroredAt    time.Time `json:"errored-at"`
+	FinishedAt   time.Time `json:"finished-at"`
+	MFAWaitingAt time.Time `json:"mfa_waiting-at"`
+	PendingAt    time.Time `json:"pending-at"`
+	QueuedAt     time.Time `json:"queued-at"`
+	RunningAt    time.Time `json:"running-at"`
+}
+
+// Read a single run by its ID.
+func (s *Plans) Read(ctx context.Context, planID string) (*Plan, error) {
+	if !validStringID(&planID) {
+		return nil, errors.New("Invalid value for plan ID")
+	}
+
+	u := fmt.Sprintf("plans/%s", url.QueryEscape(planID))
+	req, err := s.client.newRequest("GET", u, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	p := &Plan{}
+	err = s.client.do(ctx, req, p)
+	if err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+// Logs retrieves the (partial) logs of a plan.
+func (s *Plans) Logs(ctx context.Context, url string) ([]byte, error) {
```

Comment:
This seems like it should be a method hung off of the `Plan` struct. The URL is a fully-qualified URL with the protocol, host, etc. all populated in a single string. This way you never actually have to pass in a URL.

Reply:
That is true, but I like the fact that we are now reusing the existing HTTP client and logic. So (even if the call doesn't require that at the moment) it is actually making an authenticated call using the same HTTP client as was used when creating the API client.

So if anyone is using the client with a custom HTTP client, it will just work as expected (say when using PTFE). Of course we could add the client to the `Plan` struct when returning a `Plan`, but that feels a bit nasty. If we want to make the call from the `Plan` struct, it should contain all the logic to do so IMO.

Reply:
Hmm.. I just realized we could just add the HTTP client from the `Client` to the `Plan` (instead of adding the whole client). And then in the `Logs()` method we could check if we have a custom HTTP client and if not just use `http.DefaultClient` (for when someone created a `&Plan{}` manually).

So that would mitigate my earlier comment. Yet I still think it would be a better match to keep this method as is and create a new streaming version on the `Plan` struct when we actually have a streaming API. 

What do you think?

Reply:
Ah, let me explain!

So, the URL to the plan logs is a totally separate system from TFE itself - Archivist. TFE is able to generate URL's directly to the Archivist service. When requesting Archivist URL's, though, we should *not* be sending the TFE bearer token in the request. It isn't necessary, and we shouldn't send authentication tokens between services when it is not necessary for security reasons.

The log URL is meant to be called directly, with no further headers or other request details. It's fully qualified, as I noted earlier, so the client should only need to make a request to the exact URL given without any other knowledge of the Archivist service.

Fair point about embedding the client in the `Plan` - I'm ok with not doing that for the reasons you mentioned.

Reply:
I do think, however, that it might be useful to just pass a `Plan` into the `Logs` method in this case, like `Logs(ctx, plan)` (instead of just the URL). That way we can leverage additional details of the plan for things like streaming.

Reply:
Thanks for the explanation, but I knew all that already :wink: (that's why I said `even if the call doesn't require that at the moment`).

> Fair point about embedding the client in the Plan - I'm ok with not doing that for the reasons you mentioned.

But we do need the `http.Client` so that any custom client will be used instead of a default one (in case anyone passed in a custom `http.Client` with a proxy for example).

> I do think, however, that it might be useful to just pass a Plan into the Logs method in this case, like Logs(ctx, plan) (instead of just the URL). That way we can leverage additional details of the plan for things like streaming.

What additional details do we need for the stream other then the URL?

Reply:
See the other discussion... If we do have a streaming API (which I didn't knew), then I also prefer to move the `Logs()` method. So this discussion becomes irrelevant :)

Reply:
> But we do need the http.Client so that any custom client will be used instead of a default one (in case anyone passed in a custom http.Client with a proxy for example).

Right - so `Plans` will have the client embedded by way of the `client.http` and related methods you have already built. `Plan` remains a simple data-only struct. You already have it this way, and I totally get why we need the possibly custom HTTP client.

> What additional details do we need for the stream other then the URL?

I think we'll want the plan ID so that we can check if the plan has ended. Then we'll be able to either keep checking for new logs, or stop the streaming and return.